### DISCUSSION TITLE
nyagosで複数行のコマンドを実行する方法は？

### DISCUSSION THREAD

**Discussion by fushihara** [nyaosorg/nyagos#418] [Q&A] [answered]
Windowsのコマンドプロンプトではキャレット記号(^)を使ってコマンドを複数行に分割する事が出来ます

```
curl -I https://example.com ^
      https://example.com 
```
これは以下のコマンドと同じ意味です
`curl -I https://example.com https://example.com`

![image](https://user-images.githubusercontent.com/1039534/163675897-e2061ea5-77b1-4538-9fc3-1a29ddd20009.png)

これに相当する事をnyagosで行う方法はありませんでしょうか。

`Nihongo Yet Another GOing Shell 4.4.11_0-windows-amd64 by go1.17.5` では1行づつのコマンドとして扱われました。
![image](https://user-images.githubusercontent.com/1039534/163676125-b3fe6fe2-e014-4c4c-9525-beb1b82c1723.png)

( issueとは違う気がしたので、Discussionsにて起票させていただきました。 )

**Comment by hymkor** [comment, marked as answer]
どうだったかなとソースを確認したところ、行継続の仕組みは実装していなかったみたいです。

おそらくはこのあたり、行継続文字として、CMD.EXE 互換の`^` にするか、bash互換の `\` にするか迷いがあって、考えあぐねた結果、要望が来るまで保留のままにしておいたのではないかと思います。
（とはいえ、Windows だと `\` はディレクトリ区切り文字で、うっかり末尾においてしまうこともあるから、今考えれば `^` がよいかと思います）

というわけなので、今は実装されていないけれども要望をいただけるのであれば `^` による行継続を今後サポートするのは可能というのが回答になると思います。

**Comment by fushihara** [comment]
ありがとうございます。
テキストファイルに長いコマンドを記載する時、改行する表記があると見やすくなるので、実装して頂ければ嬉しいです。

要望ついでで恐縮ですが、windowsのcmd.exeで複数行のコマンドを実行した後にF7で履歴を表示する(もしくは上下キーで履歴を選択する)と
添付画像の様に各行が独立して記録されてしまいますので、ここを1件の履歴として記録して頂けたらさらに助かります。
![image](https://user-images.githubusercontent.com/1039534/163681165-c79a3da1-7c5b-4853-baa5-f6079f97bdc9.png)

**Comment by hymkor** [comment]
はい、承知です。

> 要望ついでで恐縮ですが、windowsのcmd.exeで複数行のコマンドを実行した後にF7で履歴を表示する(もしくは上下キーで履歴を選択する)と
> 添付画像の様に各行が独立して記録されてしまいますので、ここを1件の履歴として記録して頂けたらさらに助かります

はい、１件扱いの方がやりやすいといえばやりやすいです。
ただ、nyagos のコマンドライン編集は表示行は１行という体裁でしか扱えないので、ヒストリで呼び出すと複数行を連結する部分が多分 `^^J` みたいに `^` + 改行コードを制御文字入力したようなかたちになってしまいます。
そこが多分制限になると思います。

**Comment by hymkor** [comment]
- 二重引用符が閉じられていない場合
- 末尾が ^ の場合

コマンド入力を継続するようにしました。次のバイナリリリースをお待ち下さい

（入力の偶数回目は Ctrl-P によるヒストリ参照結果です）
![image](https://user-images.githubusercontent.com/3752189/163715398-59a5683f-ba8f-4b0a-94b6-d33718f87e94.png)

**Comment by hymkor** [comment]
1週間ほど試用してみたのですが「二重引用符が閉じられていない場合」を行継続にするのは今回は見送りにしたいと思います。というのも、二重引用符の単なる「閉じ忘れ」で意図せぬ行継続になって、全部入力にし直しになることが結構あるもので。

（全部打ち直しになってしまう例）  
![image](https://user-images.githubusercontent.com/3752189/164983989-70efb8ce-022e-4857-85e3-fae0eaac5a82.png)

二重引用符の中に改行を入れたい時は、こちらも `^` を使うことにしたいと思います。

![image](https://user-images.githubusercontent.com/3752189/164984204-1606bb5b-dee6-4edb-bf99-cb79cb16cd23.png)

![image](https://user-images.githubusercontent.com/3752189/164984231-31166c1b-e4bc-4906-b6e4-bcca8d630176.png)

**Comment by hymkor** [comment]
> -  末尾が ^ の場合
>
> コマンド入力を継続するようにしました。次のバイナリリリースをお待ち下さい

報告が漏れていました。[4.4.12_0](https://github.com/nyaosorg/nyagos/releases/tag/4.4.12_0) にて実装版が公開されています。
（2023.10.17現在の最新は [4.4.14_0](https://github.com/nyaosorg/nyagos/releases/tag/4.4.14_0) なので、だいぶと間があいてしまいましたが…

本件、クローズさせていただきます。
何か問題などありましたら、再オープンください。
